### ISSUE TITLE
Support for CI integration

### ISSUE THREAD

**Issue by johbo** [compilerla/conventional-pre-commit#41]
I've just worked myself through a question how to best integrate the checks into a ci pipeline. Would like to share my experience from the approach and also trigger a question if it's possible to implement support for such use cases.

In our team the context is as follows:

* We work locally and some have the checks based on `pre-commit` also enabled locally.
* Then work is pushed into a branch and a Merge Request / Pull Request is opened based on the branch.
* On this Merge Request we wanted to trigger the same linting checks, so that we never have to worry about commit message structure, even if a colleague decided to not enable `pre-commit` locally. Latest on the Merge Request the automated feedback would be available.

While building this (in our case based on gitlab ci) I found it a particular challenge that I need to extract every commit message manually and then run `pre-commit` to have the message checked. I've reached a working state for us, and still I wonder if it wouldn't be a good improvement to be able to run a check by passing the hashes of a starting and ending commit.

As pseudo code it could look something like the following:

```
pre-commit --from-ref=XY --tro-ref=YZ --hook-stage=commit-msg
```

The below example shows how I've done it currently and probably helps to understand how I got to the question/suggestion:

```shell
      any_failure=0
      tmp_file="$(mktemp --tmpdir=.)"
      for x in $(git log --format="%H" "${CI_MERGE_REQUEST_DIFF_BASE_SHA}..${CI_COMMIT_SHA}")
      do
          echo "Checking commit message for $x"
          git log --format="%B" -n 1 "$x" > "$tmp_file"
          pre-commit run --hook-stage=commit-msg --commit-msg-filename="$tmp_file" \
              conventional-pre-commit || any_failure=1
      done
      rm "$tmp_file"
      if [ "$any_failure" -ne 0 ]
      then
          exit 1
      fi
```

**Comment by thekaveman** [comment]
@johbo Thank you for the question. Indeed, your use-case is one that has been discussed here before: https://github.com/compilerla/conventional-pre-commit/issues/9#issuecomment-966490052. Please take a look at this thread for more context.

If I am understanding this proposal / question correctly:

> I wonder if it wouldn't be a good improvement to be able to run a check by passing the hashes of a starting and ending commit

(and based on the pseudocode), it seems like the change you request would actually be in [`pre-commit`](https://github.com/pre-commit/pre-commit) itself, to accept a range of commits to run (arbitrary hooks/stages) against?

The TLDR; answer for here is we won't be making any changes to `conventional-pre-commit` to support CI, because this is fundamentally incompatible with the git `commit-msg` hook/phase.

**Comment by ssbarnea** [comment]
I do think that it would greatly useful to also provide a solution for those that want to run this check after the commit was already created, in `pre-commit` or `pre-push`.

I might not necessary be the same hook definition but it should be defined here too. That is of high importance not only for CI integration but also for huge number of pre-commit/prek users which might not install hooks but run run `... run -a` instead manually.
